Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -106,6 +119,55 @@ protected override BoolResult InitializeCore(OperationContext context)
             return result;
         }
 
+        /// <inheritdoc />
+        public override void SetDatabaseMode(bool isDatabaseWriteable)
+        {
+            if (IsGarbageCollectionEnabled != isDatabaseWriteable)
+            {
+                var nextCompactionTimeSpan = isDatabaseWriteable ? _configuration.FullRangeCompactionInterval : Timeout.InfiniteTimeSpan;
+                _compactionTimer?.Change(nextCompactionTimeSpan, Timeout.InfiniteTimeSpan);
+            }
+
+            base.SetDatabaseMode(isDatabaseWriteable);
+        }
+
+        private void FullRangeCompaction(OperationContext context)
+        {
+            lock (ShutdownLock)
+            {
+                if (ShutdownStarted)
+                {
+                    return;
+                }
+            }
+
+            // We can safely ignore the result, as any relevant errors will be logged by the inner operations
+            _keyValueStore.Use(store =>
+            {
+                foreach (var columnFamilyName in new[] { "default", nameof(Columns.ClusterState), nameof(Columns.Metadata) })
+                {
+                    var result = context.PerformOperation(Tracer, () =>
+                    {
+                        store.CompactRange((byte[])null, null, columnFamilyName: columnFamilyName);
+                        return BoolResult.Success;
+                    }, extraStartMessage: $"ColumnFamily={columnFamilyName}");
+
+                    if (!result.Succeeded)
+                    {
+                        break;
+                    }
+                }
+            }).ToBoolResult().IgnoreFailure();
+
+            lock (ShutdownLock)
+            {
+                if (!ShutdownStarted)
+                {
+                    _compactionTimer?.Change(_configuration.FullRangeCompactionInterval, Timeout.InfiniteTimeSpan);

```

Comment:
Timer restart should be inside try/finally block.

We don't want to stop the collection if the exception will occur here.

Reply:
According to documentation, Timer.Change can only throw if the timer has already been disposed of or if parameters are wrong. The first condition should never happen due to the check for shutdown, the others would cause a failure much earlier.

---
In reply to: [313645457](https://github.com/microsoft/BuildXL/pull/731#discussion_r313645457) [](ancestors = 313645457)

Reply:
The problem that I see is that `_keyvaluestore.Use` can throw, or the compation itself

---
In reply to: [313647176](https://github.com/microsoft/BuildXL/pull/731#discussion_r313647176) [](ancestors = 313647176,313645457)

Reply:
if _keyValueStore.Use (which is just taking a lock) throws, everything will fail, not just this function, so there's nothing we can do.

If compaction fails, the inner operation ("result") will fail, which means we will trace and break.

---
In reply to: [313649456](https://github.com/microsoft/BuildXL/pull/731#discussion_r313649456) [](ancestors = 313649456,313647176,313645457)